tests/testthat/test-utils.R

```diff
@@ -415,31 +415,63 @@ test_that("bin() returns appropriate bins for dates", {
   viewport <- list('xMin'=min(testDF$entity.dateA), 'xMax'=max(testDF$entity.dateA))
   dt <- testDF
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), 'day', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), 'day', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-12-01 - 1999-12-02")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), '4 day', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), '4 day', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-11-29 - 1999-12-03")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), 'week', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), 'week', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-11-29 - 1999-12-06")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), '2 week', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), '2 week', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-11-29 - 1999-12-13")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), 'month', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), 'month', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-12-01 - 2000-01-01")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), '2 month', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), '2 month', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-11-01 - 2000-01-01")   
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), 'year', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), 'year', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-01-01 - 2000-01-01")
 
-  dt$dateBin <- bin(as.Date(testDF$entity.dateA), '2 year', viewport)
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), '2 year', viewport)
   expect_equal(unique(dt$dateBin[dt$entity.dateA == '1999-12-01']), "1999-01-01 - 2001-01-01")
 })
 
+test_that("bin() returns complete list of possible bins as levels when asked for a factor.", {
+  viewport <- list('xMin'=min(testDF$entity.dateA), 'xMax'=max(testDF$entity.dateA))
+  dt <- testDF
+
+  dt$dateBin <- bin(as.Date(dt$entity.dateA), 'day', viewport, stringsAsFactors=T)
+  expect_equal(uniqueN(dt$dateBin), 393)
+  expect_equal(uniqueN(levels(dt$dateBin)), 3986)
+
+  viewport <- list('xMin'=min(dt$entity.contA), 'xMax'=max(dt$entity.contA))
+  binWidth <- plot.data::findBinWidth(dt$entity.contA) 
+  dt$contBin <- bin(dt$entity.contA, binWidth, viewport, stringsAsFactors=T)
+  expect_equal(uniqueN(dt$contBin), 16)
+  expect_equal(uniqueN(levels(dt$contBin)), 17)
```

Comment:
why do we get an extra bin at the upper end here? The max of `dt$entity.contA` is 14.36 which is the same as the viewport max, but the extra bin starts at 14.4.

Reply:
I'd guess same reason Bob was asking about integer bin labels for non integer bin widths.. the last bin doesn't actually start at 14.4, that's just a rounded label. Haven't confirmed yet though